### ISSUE TITLE
sqladmin: SslCerts.CreateEphemeral returning 404 in v0.14.0 and working in v0.13.0

### ISSUE THREAD

**Issue by srikrsna** [googleapis/google-api-go-client#427]
We use https://github.com/GoogleCloudPlatform/cloudsql-proxy to connect to our database. It internally calls `sqladmin.SslCerts.CreateEphemeral` to create an ephemeral certificate for the encryption. With the same configuration this method is behaving differently between v0.14.0 and v0.13.0

Basically it returns a 404 in the case of v0.14 and in v0.13 it is running without any errors. I did not open an issue in that repo as there were no changes there.

Could you please look into this?

**Comment by jadekler** [comment]
@kurtisvg Any idea what this is about?

**Comment by kurtisvg** [comment]
@srikrsna - Can you provide more details logs and if possible the configuration you are using? Please censor any identifying information. 

Are you using the proxy as a binary or just the dialer portion of it? Is the `v.0.14.0` a reference to the library version? Cloud SQL proxy doesn't have releases matching those version numbers.

**Comment by jadekler** [comment]
> Is the v.0.14.0 a reference to the library version?

I believe that is https://github.com/googleapis/google-api-go-client/releases/tag/v0.14.0.

**Comment by srikrsna** [comment]
> Are you using the proxy as a binary or just the dialer portion of it? Is the `v.0.14.0` a reference to the library version?

We are using CloudSQL proxy library in exactly the same way as shown [here](https://github.com/GoogleCloudPlatform/cloudsql-proxy/blob/master/proxy/dialers/postgres/hook_test.go). 

Cloud SQL Proxy (Binary) is locked to v0.9.0 of this library and hence that is working. But since we are also using other libraries (stackdriver) that depend on the master branch of this we received this error. For now we have locked the version to v0.13 to get things working.

Here's the error we get when we try to connect using v0.14.0

`ensure that the account has access to \"<insatnce>\" (and make sure there's no typo in that name). Error during createEphemeral for <instance>: googleapi: got HTTP response code 404 with body: Not Found`

The same configuration works if lock the version to v0.13.0.

**Comment by srikrsna** [comment]
I am able to replicate the issue by cloning the cloudsql-proxy and upgrading the version from v.0.9 to v0.14. I've tested it on two different instances in different projects.

**Comment by srikrsna** [comment]
I'm not sure what the fix is but I think I found the issue.

The final request that's going out is 
`https://www.googleapis.com/sql/v1beta4/sql/v1beta4/projects/{project}/instances/{instance}/createEphemeral?alt=json&prettyPrint=false`
If you notice `sql/v1beta4` is appended twice. This definitely happened due to the changes after v0.13.0 to this packaged. 

Basically in one of the commits base url was set to empty string and `sql/v1beta4` was prepended to all API calls. The `certs` package in `cloudsql-proxy` sets its own base url. When the new commits to this package were added it caused the final call to contain 2 `sql/v1beta4`'s.

[Here's](https://github.com/GoogleCloudPlatform/cloudsql-proxy/blob/46c5533ff5bad7ea6a5e4966a670249d299f5880/proxy/certs/certs.go#L88) the line in `cloudsql-proxy`

@jadekler @kurtisvg Hope this helps. Let me know if you need anything else.

**Comment by shollyman** [comment]
From the diff of the sqladmin discovery document, there's quite a few things that changed in the sqladmin v1beta4 API, which were picked up as v0.14.0 when I ran the regen process last week.

* API moved to an api-specific host (www.googleapis.com -> sqladmin.googleapis.com)
* The basepath was truncated and propagated to the individual path references in the discovery doc
* Support for media and proto alt responses was removed.
* for the CreateEphemeral method, it seems there's a new "parent" field that's been added

This suggests that the calling code (e.g. in cloudsql-proxy) may need updates to reflect these changes in the sqladmin API, or that the backend team has further issues with their discovery document that need addressed.

**Comment by kurtisvg** [comment]
@srikrsna @shollyman - Thanks for the additional info, this does indeed help.

I'll reach out to the product team and confirm these changes are intentional, and if so update the Cloud SQL proxy/dialer as necessary.

**Comment by waelmas** [comment]
I reproduced the same issue after coming across this issue 3 days ago. _(Before this discussion started so I was trying to make sure it was not a one-time thing)_

Reproduced using the default [cloudsqlpostgres](https://github.com/GoogleCloudPlatform/cloudsql-proxy/blob/master/proxy/dialers/postgres/hook_test.go) dialer. It seems to me that as mentioned by others, the recent release of googleapis v14 caused compatibility issues with the dialer.
The error seems to be a result of the dialer not connecting at all and thus returning a 404 error code. _(even if you try providing an invalid host to the dialer the same exact error occurs)_

[Configuration example](https://cloud.google.com/sql/docs/postgres/connect-external-app#go-proxy-cp)

Using go1.13.4 linux/amd64 and a PostgreSQL 9.6 Cloud SQL instance.

Error code:
`“ensure that the account has access to "[INSTANCE_CONNECTION_NAME]" (and make sure there's no typo in that name). Error during createEphemeral for [INSTANCE_CONNECTION_NAME]: googleapi: got HTTP response code 404`

However using the [companion process configuration](https://cloud.google.com/sql/docs/postgres/connect-external-app#go-proxy-cp) everything works fine. (I have compiled the Cloud SQL Proxy from source)

**Comment by srikrsna** [comment]
> * The basepath was truncated and propagated to the individual path references in the discovery doc

I think this should be considered a breaking change considering the exposed API by this package to change the base path.

**Comment by colega** [comment]
This appears to be fixed by changing the hardcoded certs source path in `cloudsql-proxy`: https://github.com/GoogleCloudPlatform/cloudsql-proxy/commit/eb5e09ef70a515db75cc59caaee50d29638ad5db#diff-2e191ab6bb0c70e18327a004be427a10R75

**Comment by rick-pri** [comment]
> This appears to be fixed by changing the hardcoded certs source path in `cloudsql-proxy`: [GoogleCloudPlatform/cloudsql-proxy@eb5e09e#diff-2e191ab6bb0c70e18327a004be427a10R75](https://github.com/GoogleCloudPlatform/cloudsql-proxy/commit/eb5e09ef70a515db75cc59caaee50d29638ad5db#diff-2e191ab6bb0c70e18327a004be427a10R75)

This is not the case as far as I can tell.  Running through the following to compile the proxy:

```
setenv GO111MODULE auto
go get github.com/GoogleCloudPlatform/cloudsql-proxy/cmd/cloud_sql_proxy
```

and attempting connecting to a Postgres 11 CloudSQL instance with a cloudsql proxy compiled from master still results in:

```
  <p><b>404.</b> <ins>That’s an error.</ins>
  <p>The requested URL <code>/sql/v1beta4/sql/v1beta4/projects/<proj-name>/instances/dm-cloud-sql-postgres?alt=json&amp;prettyPrint=false</code> was not found on this server.  <ins>That’s all we know.</ins>
```

Of course, I can check the repo out in `~/go/src/github.com/GoogleCloudPlatform/cloudsql-proxy/cmd/cloud_sql_proxy` at the 1.16 tag and then building the binary works just fine.

**Comment by kurtisvg** [comment]
This issue is fixed in https://github.com/GoogleCloudPlatform/cloudsql-proxy/commit/eb5e09ef70a515db75cc59caaee50d29638ad5db.

Please make sure you are either using a version of GoogleCloudPlatform/cloudsql-proxy past that commit, or that the version of `google-api-go-client` is v0.13.0 or before.

If you are still seeing the issue with `/sql/v1beta4/sql/v1beta4` it means you there is a mismatch. If you are sure that you are on compatible versions and still seeing the error, please open an issue on GoogleCloudPlatform/cloudsql-proxy (since the issue is actually with the proxy, and not `google-api-go-client`).

**Comment by dim** [comment]
@kurtisvg would it be possible to tag a new release?

**Comment by rick-pri** [comment]
> This issue is fixed in [GoogleCloudPlatform/cloudsql-proxy@eb5e09e](https://github.com/GoogleCloudPlatform/cloudsql-proxy/commit/eb5e09ef70a515db75cc59caaee50d29638ad5db).
> 

As far as I can tell this was merged into master so when I did the `go get` I should have compiled from master and had that change in it, which was why I thought that the issue was here.  It seems that the issue is that the CloudSQL Proxy is pulling in this repo at version `v0.15.0`.

> If you are still seeing the issue with `/sql/v1beta4/sql/v1beta4` it means you there is a mismatch. If you are sure that you are on compatible versions and still seeing the error, please open an issue on GoogleCloudPlatform/cloudsql-proxy (since the issue is actually with the proxy, and not `google-api-go-client`).

I'll do this and reference this issue.
